Update documentation for step-up authentication (#1428)
keycloak/keycloak-documentation: server_admin/topics/authentication/flows.adoc

. Click *Config*.
. Enter `Level 1` as an alias.
. Enter `1` for the Level of Authentication (LoA).

> . Set Max Age to *36000*. This is in seconds and it is equivalent of 10 hours, which is default `SSO Session Max` timeout set in the realm.

Reviewer: always add a noun after this  --- this value
change "of" to "to"
add "the" before "default"

>   It means that when user authenticates with this level, the further SSO logins will be able to re-use this level and user will not need to authenticate

Reviewer 2: `It means that when a user authenticates with this level,` might be good.

Reviewer 2: `and the user will not need to authenticate` might be good.

Author: Applied, Thanks!

>   It means that when a user authenticates with this level, the further SSO logins will be able to re-use this level and the user will not need to authenticate

Reviewer: Consider this phrasing
As a result, when a user authenticates with this level, subsequent SSO logins can re-use this level and the user does  not need to authenticate

Author: Applied, Thanks

>   with this level until the end of the user session (10 hours).

Reviewer: Change " session (10 hours)
to
session, which is 10 hours

. Click *Config*.
. Enter `Level 2` as an alias.
. Enter `2` for the Level of Authentication (LoA).

> . Set Max Age to *0*. This means that when user authenticates, this level is valid just for the current authentication, but not any

Reviewer 2: `This means that when a user authenticates,` might be good.

Reviewer 2: `So the user will always need to authenticate again with this level when this level is requested.` might be good.

Author: Applied

> . Set Max Age to *0*. This means that when a user authenticates, this level is valid just for the current authentication, but not any

Reviewer: Change "This means that" to "As a result,"

>   further SSO authentications. So the user will always need to authenticate again with this level when this level is requested.

Reviewer: Change "further" to "subsequent"

image:images/authentication-step-up-flow.png[]

.Request a certain authentication level

> To use the step-up mechanism, you have to specify a requested level of authentication (LoA) in your authentication request. The `claims` parameter is used for this purpose:

Reviewer: remove "have to"  -- adds no meaning

        }
----

The {project_name} javascript adapter has support for easy construct of this JSON and sending it in the login request.
See link:{adapterguide_link_js_adapter}[Javascript adapter documentation] for more details.

> It is also possible to use simpler parameter `acr_values` instead of `claims` parameter to request particular levels as non-essential. This is mentioned

Reviewer: Change "It is also possible to"
to
"You can also"

in the OIDC specification.

> There is also possibility to configure default level for the particular client, which is used when parameter `acr_values` or parameter `claims` with the `acr` claim is not present.

Reviewer: Change to 
"You can also configure the default..."
add "the" before "parameter" (both times)

> For further details, see <<_mapping-acr-to-loa-client,Client ACR configuration>>).

Reviewer 3: ```suggestion
For further details, see <<_mapping-acr-to-loa-client,Client ACR configuration>>.
```

> NOTE: To request the acr_values as text (such as `gold`) instead of a numeric value, you have to configure the mapping between the ACR and the LoA.

Reviewer: remove "have to" -- adds no meaning

It is possible to configure it at the realm level (recommended) or at the client level. For configuration see <<_mapping-acr-to-loa-realm,ACR to LoA Mapping>>.

For more details see the https://openid.net/specs/openid-connect-core-1_0.html#acrSemantics[official OIDC specification].

*Flow logic*

The logic for the previous configured authentication flow is as follows: +

> If a client request a high authentication level, meaning Level of Authentication 2 (LoA 2), a user has to perform full 2-factor authentication: Username/Password + OTP.

Reviewer 3: ```suggestion
If a client requests a high authentication level, meaning Level of Authentication 2 (LoA 2), a user has to perform full 2-factor authentication: Username/Password + OTP.
```

> However, if a user already has a session in Keycloak, that was logged in with username and password (LoA 1), the user is only asked for the second authentication factor (OTP).

Reviewer 3: ```suggestion
However, if the user has been already authenticated by the first authentication factor (Username/Password - LoA 1), the user is only asked for the second authentication factor (OTP - LoA 2).
```
It's maybe better, right?

> The option *Max Age* in the condition determines for how long (how much seconds) is the subsequent authentication level valid. This helps to decide

Reviewer 2: `The option *Max Age* in the condition determines how long (how much seconds) the subsequent authentication level is valid.` might be good.

Author: Applied

> The option *Max Age* in the condition determines how long (how many seconds) the subsequent authentication level is valid. This helps to decide

Reviewer: This setting helps to decide...

> wheather the user will be asked to present the authentication factor again during a subsequent authentication. If the particular level X is requested

Reviewer 2: typos: `whether` might be appropriate.

Author: Applied

by the `claims` or `acr_values` parameter and the user has been already authenticated with level X, but it is expired (for example max age is configured to 300 and user authenticated before 310 seconds)

> then the user will be asked to re-authenticate again with the particular level. However, if the level is not yet expired, user will be automatically

Reviewer: add "the" before "user will be automatically..."

> The option *Max Age* in the condition determines how long (how much seconds) the subsequent authentication level is valid. This helps to decide
> whether the user will be asked to present the authentication factor again during a subsequent authentication. If the particular level X is requested
> by the `claims` or `acr_values` parameter and user already authenticated with level X, but it is expired (for example max age is configured to 300 and user authenticated before 310 seconds)
> then the user will be asked to re-authenticate again with the particular level. However if the level is not yet expired, user will be automatically
> considered as authenticated with that level.

Reviewer 3: ```suggestion
The option *Max Age* in the condition determines how long (how many seconds) the subsequent authentication level is valid. This helps to decide
whether the user will be asked to present the authentication factor again during a subsequent authentication. If the particular level X is requested
by the `claims` or `acr_values` parameter and the user has been already authenticated with level X, but it is expired (for example max age is configured to 300 and user authenticated before 310 seconds)
then the user will be asked to re-authenticate again with the particular level. However, if the level is not yet expired, user will be automatically
considered as authenticated with that level.
```

Using *Max Age* with value 0 means, that particular level is valid just for this single authentication. Hence every re-authentication requesting that level

> will need to authenticate again with that level. This is useful for operations that require higher security in the application (e.g. send payment) and always require authentication

Reviewer: add "the" before "value 0 means"

with the specific level.

> WARNING: Note that parameters like `claims` or `acr_values` might be changed by the user in the URL when the login request is sent from the client to the {project_name}.

Reviewer 2: `when the login request is sent from the client to the {project_name} via the user's browser.` might be a precise expression.

Author: Applied

> WARNING: Note that parameters like `claims` or `acr_values` might be changed by the user in the URL when the login request is sent from the client to the {project_name} via the user's browser.

Reviewer: change "like" to "such as"

> Unless client uses PAR (Pushed authorization request) or other mechanisms to rewrite URL. Hence after the authentication, clients are encouraged to check the ID Token to

Reviewer 2: `Unless a client uses PAR (Pushed Authorization Request), a resource object, or other mechanisms to prevent rewriting the URL, clients are encouraged to` might be good.

Author: I guess you meant "a request object" ? I've applied with the "a request object" instead of "a resource object" .

> This can be mitigated if the client uses PAR (Pushed authorization request), a request object, or other mechanisms, which prevents the user from rewriting the parameters in the URL.

Reviewer: This situation..
Change "mechanisms, which prevents" to "mechanisms that prevents"

> doublecheck that `acr` in the token corresponds to the expected level.

Reviewer 2: It might be good that L379 and L380 be one sentence.

Author: I've slightly rewrote this. However I did not merge into one sentence. I think it is always better that client will check the content of "acr" claim (even if client uses PAR).

Reviewer 2: @mposolda I see.

> In case that there is no explicit level requested by parameters, the {project_name} will require the

Reviewer 2: From L382 to L385, line feeds are not appropriate.

Author: Fixed

> WARNING: Note that parameters like `claims` or `acr_values` might be changed by the user in the URL when the login request is sent from the client to the {project_name} via the user's browser.
> This can be mitigated if client uses PAR (Pushed authorization request), a request object, or other mechanisms, which prevents the user from rewrite the parameters in the URL.
> Hence after the authentication, clients are encouraged to check the ID Token to doublecheck that `acr` in the token corresponds to the expected level.

Reviewer 3: ```suggestion
WARNING: Note that parameters like `claims` or `acr_values` might be changed by the user in the URL when the login request is sent from the client to the {project_name} via the user's browser.
This can be mitigated if the client uses PAR (Pushed authorization request), a request object, or other mechanisms, which prevents the user from rewriting the parameters in the URL.
Hence after the authentication, clients are encouraged to check the ID Token to doublecheck that `acr` in the token corresponds to the expected level.
```

> authentication with the first LoA condition found in the authentication flow (Username/Password in the example above). When user was already

Reviewer 2: `When a user was already` might be good.

Author: Applied

> In case that there is no explicit level requested by parameters, the {project_name} will require the authentication with the first LoA

Reviewer: If no explicit level is requested by parameters..

> authenticated with that level and that level expired, user is not required to re-authenticate, but `acr` in the token will have value 0. This is

Reviewer 2: `the user is not required` might be good.

Author: Applied

> condition found in the authentication flow (Username/Password in the example above). When a user was already authenticated with that level

Reviewer: Change: flow (Username/Password in the example above)
TO
flow, such as in the preceding Username/Password example

> and that level expired, the user is not required to re-authenticate, but `acr` in the token will have value 0. This is considered as authentication

Reviewer: will have the value 0
This result is considered..

based solely on `long-lived browser cookie` as mentioned in the section 2 of OIDC Core 1.0 specification.

NOTE: A conflict situation may arise when an admin specifies several flows, sets different LoA levels to each, and assigns the flows to different clients. However, the rule is always the same: if a user has a certain level, it needs only have that level to connect to a client. It's up to the admin to make sure that the LoA is coherent.

*Example scenario*

. Max Age is configured as 300 seconds for level 1 condition.

> . Login request is sent without requesting any acr. Level 1 will be used and user needs to authenticate with username and password. Token will have `acr=1`.

Reviewer: Add "the" before "user" and before "Token"    The token...

> . Another login request is sent after 100 seconds. User automatically authenticated due the SSO and token will return `acr=1`.

Reviewer: Change "User" to "The user is.."
add "to" after "due"

> . Another login request is sent after another 201 seconds (301 seconds since authentication in point 2). User automatically authenticated due the SSO, but token will return `acr=0` due the level 1 is considered expired.

Reviewer: Add "the" before "user"  -- the user is...
add "to" after "due"
Add "the"  before "token"

> . Another login request is sent, but now it will explicitly request ACR of level 1 in the `claims` parameter. User will be asked to re-authenticate with username/password

Reviewer 2: It might be good that L395 and L396 be one sentence.

Author: Hmm.. From what I heard somewhere, it can be better to use more sentences (if possible) rather than one big sentence as more sentences are usually easier for readers to follow. So I am keeping as it is for now. If you strongly prefer to change this, please let me know how you suggest those sentences to exactly look like.

Reviewer 2: Sorry. What I wanted to say is to remove linefeeds among L395-L396. Not write them as one sentence.

Author: Replied in the comment https://github.com/keycloak/keycloak-documentation/pull/1428#discussion_r824468705 . We can keep the conversation there related to line feeds.

Reviewer 2: Yes, I agree.

  and then `acr=1` will be returned in the token.

*ACR claim in the token*

> ACR claim is added to the token by the `acr loa level` protocol mapper defined in the `acr` client scope. This client scope is the realm default client scope

Reviewer 2: It might be good that L400 and L401 might be one sentence.

Author: Same as above

Reviewer 2: Sorry. What I wanted to say is to remove linefeeds among L400-L401. Not write them as one sentence.

Author: Replied in the comment https://github.com/keycloak/keycloak-documentation/pull/1428#discussion_r824468705 . We can keep the conversation there related to line feeds.

Reviewer 2: Yes, I agree.

and hence will be added to all newly created clients in the realm.

In case you do not want `acr` claim inside tokens or you need some custom logic for adding it, you can remove the client scope from your client.

> Note when login request initiates request with `claims` parameter requesting `acr` as `essential` claim, then {project_name} will always return

Reviewer 2: It might be good that L405-L407 might be one sentence.

Author: Same as above

Reviewer 2: Sorry. What I wanted to say is to remove linefeeds among L405-L407. Not write them as one sentence.

Author: Replied in the comment https://github.com/keycloak/keycloak-documentation/pull/1428#discussion_r824468705 . We can keep the conversation there related to line feeds.

Reviewer 2: Yes, I agree.

> Note when login request initiates request with `claims` parameter requesting `acr` as `essential` claim, then {project_name} will always return

Reviewer: Add "the" before "login"
Add "a" before "request"
add "the" before 'claims` paremeter

> one of the specified levels. If it is not able to return one of the specified levels (For example if requested level is unknown or bigger than configured conditions

Reviewer: Add "the" before "requested"
